process/block/preprocess/miniBlocksCompaction.go

```diff
@@ -0,0 +1,289 @@
+
+	return mergedMiniBlocks
+}
+
+func (mbc *miniBlocksCompaction) haveEnoughGasToMerge(
+	destMiniBlock *block.MiniBlock,
+	srcMiniBlock *block.MiniBlock,
+) bool {
+
+	gasUsedInDestMiniBlock, err := mbc.getGasUsedInMiniBlock(destMiniBlock)
+	if err != nil {
+		log.Info(err.Error())
+		return false
+	}
+
+	gasUsedInSrcMiniBlock, err := mbc.getGasUsedInMiniBlock(srcMiniBlock)
+	if err != nil {
+		log.Info(err.Error())
+		return false
+	}
+
+	haveEnoughGasToMerge := gasUsedInDestMiniBlock+gasUsedInSrcMiniBlock <= process.MaxGasLimitPerMiniBlock
+
+	return haveEnoughGasToMerge
+}
+
+func (mbc *miniBlocksCompaction) getGasUsedInMiniBlock(miniBlock *block.MiniBlock) (uint64, error) {
+	gasUsedInMiniBlock := uint64(0)
+	for _, txHash := range miniBlock.TxHashes {
+		tx, ok := mbc.mapHashesAndTxs[string(txHash)]
+		if !ok {
+			return 0, process.ErrMissingTransaction
+		}
+
+		txGasLimit := mbc.economicsFee.ComputeGasLimit(tx)
+		if isSmartContractAddress(tx.GetRecvAddress()) {
+			txGasLimit = tx.GetGasLimit()
+		}
+
+		gasUsedInMiniBlock += txGasLimit
+	}
+
+	return gasUsedInMiniBlock, nil
+}
+
+// Expand method tries to expand the given mini blocks to their initial state before compaction
+func (mbc *miniBlocksCompaction) Expand(
+	miniBlocks block.MiniBlockSlice,
+	mapHashesAndTxs map[string]data.TransactionHandler,
+) (block.MiniBlockSlice, error) {
+
+	mbc.mutMiniBlocksCompaction.Lock()
+	defer mbc.mutMiniBlocksCompaction.Unlock()
+
+	mbc.mapHashesAndTxs = mapHashesAndTxs
+	mbc.mapSenderNonce = make(map[string]uint64)
+	mbc.mapUnallocatedTxsHashes = make(map[string]struct{})
+
+	expandedMiniBlocks := make(block.MiniBlockSlice, 0)
+	miniBlocksToExpand := make(block.MiniBlockSlice, 0)
+
+	for _, miniBlock := range miniBlocks {
+		if miniBlock.SenderShardID == mbc.shardCoordinator.SelfId() {
+			miniBlocksToExpand = append(miniBlocksToExpand, miniBlock)
+			continue
+		}
+
+		expandedMiniBlocks = append(expandedMiniBlocks, miniBlock)
+	}
+
+	if len(miniBlocksToExpand) > 0 {
+		expandedMiniBlocksFromMe, err := mbc.expandMiniBlocks(miniBlocksToExpand)
+		if err != nil {
+			return nil, err
+		}
+
+		expandedMiniBlocks = append(expandedMiniBlocks, expandedMiniBlocksFromMe...)
+	}
+
+	if len(miniBlocks) < len(expandedMiniBlocks) {
+		log.Info(fmt.Sprintf("expanded from %d miniblocks to %d miniblocks\n",
+			len(miniBlocks), len(expandedMiniBlocks)))
+	}
+
+	return expandedMiniBlocks, nil
+}
+
+func (mbc *miniBlocksCompaction) expandMiniBlocks(miniBlocks block.MiniBlockSlice) (block.MiniBlockSlice, error) {
+	for _, miniBlock := range miniBlocks {
+		for _, txHash := range miniBlock.TxHashes {
+			tx, ok := mbc.mapHashesAndTxs[string(txHash)]
+			if !ok {
+				return nil, process.ErrMissingTransaction
+			}
+
+			nonce, ok := mbc.mapSenderNonce[string(tx.GetSndAddress())]
+			if !ok || nonce > tx.GetNonce() {
+				mbc.mapSenderNonce[string(tx.GetSndAddress())] = tx.GetNonce()
+			}
+
+			mbc.mapUnallocatedTxsHashes[string(txHash)] = struct{}{}
```

Comment:
so initially every transaction is unallocated ? in terms of expansion

Reply:
yes